com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,166 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
+            {
+                if (Entries[i].Fresh && Entries[i].Key.TickWritten > 0)
+                {
+                    var networkVariable = FindNetworkVar(Entries[i].Key);
+
+                    if (networkVariable != null)
+                    {
+                        Stream.Seek(Entries[i].Position, SeekOrigin.Begin);
+
+                        // todo: consider refactoring out in its own function to accomodate
+                        // other ways to (de)serialize
+                        // todo --M1--
+                        // Review whether tick still belong in netvar or in the snapshot table.
+                        networkVariable.ReadDelta(Stream, m_NetworkManager.IsServer);
+                    }
+                }
+
+                Entries[i].Fresh = false;
+            }
+        }
+
+        /// <summary>
+        /// Read the snapshot index from a buffer
+        /// Stores the entry. Allocates memory if needed. The actual buffer will be read later
+        /// </summary>
+        /// <param name="reader">The reader to read the index from</param>
+        internal void ReadIndex(NetworkReader reader)
+        {
+            Entry entry;
+            short entries = reader.ReadInt16();
+
+            for (var i = 0; i < entries; i++)
+            {
+                entry = ReadEntry(reader);
+                entry.Fresh = true;
+
+                int pos = Find(entry.Key);
+                if (pos == Entry.NotFound)
+                {
+                    pos = AddEntry(entry.Key);
+                }
+
+                // if we need to allocate more memory (the variable grew in size)
+                if (Entries[pos].Length < entry.Length)
+                {
+                    AllocateEntry(ref entry, pos, entry.Length);
+                }
+
+                Entries[pos] = entry;
+            }
+        }
+
+        /// <summary>
+        /// Helper function to find the NetworkVariable object from a key
+        /// This will look into all spawned objects
+        /// </summary>
+        /// <param name="key">The key to search for</param>
+        private INetworkVariable FindNetworkVar(VariableKey key)
+        {
+            var spawnedObjects = m_NetworkManager.SpawnManager.SpawnedObjects;
+
+            if (spawnedObjects.ContainsKey(key.NetworkObjectId))
+            {
+                var behaviour = spawnedObjects[key.NetworkObjectId]
+                    .GetNetworkBehaviourAtOrderIndex(key.BehaviourIndex);
+                return behaviour.NetworkVariableFields[key.VariableIndex];
+            }
+
+            return null;
         }
     }
 
     public class SnapshotSystem : INetworkUpdateSystem, IDisposable
     {
-        private Snapshot m_Snapshot = new Snapshot();
-        private Snapshot m_ReceivedSnapshot = new Snapshot();
         private NetworkManager m_NetworkManager = NetworkManager.Singleton;
+        private Snapshot m_Snapshot = new Snapshot(NetworkManager.Singleton, false);
```

Comment:
perhaps rename to `m_ServerSnapshot?`

Reply:
Good point. But the `spawn` version is now using a single snapshot, so the naming might not be super important. :-)

Also, when it doesn't come with extra cost, I try to abstract away the difference between server and client. Maybe one day we'll end up with hybrid combination: partial meshes, connection between hosts, etc... Then, the less "server"- and "client"-specific code we have, the most flexible it should be.